Bugfix/issue 487 hypergeometric functions

sakrejda opened the pull request:
#### Submisison Checklist

- [x] Run unit tests: `./runTests.py test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary:

There are currently bugs in F32 and ... I think it's grad_F32, and some potential bugs in grad_2F1 that I haven't demonstrated yet (it _can_ terminate at an arbitrary point).  See issue #487.  

#### Intended Effect:

Fix bug in F32 and grad_F32, add some doc and tests.

#### How to Verify:

Run new tests versus `develop` to see them fail, they pass with this branch.  On `develop` at least one of the tests hangs on my machine.  Test values come from Mathematica and R's hypergeom package.  The tests that fail against develop are:

```
cd ${STAN-MATH}
git checkout develop
git pull origin develop
git checkout bugfix/issue-487-hypergeometric-functions -- test/unit/math/prim/scal/fun/F32_test.cpp
make test/unit/math/prim/scal/fun/F32_test
./test/unit/math/prim/scal/fun/F32_test
git checkout bugfix/issue-487-hypergeometric-functions -- test/unit/math/prim/scal/prob/beta_binomial_cdf_log_test.cpp
make tests/unit/math/prim/scal/prob/beta_binomial_cdf_log_test
./test/unit/math/prim/scal/prob/beta_binomial_cdf_log_test
```

#### Side Effects:

betabinomial_cdf is now more correct and has a test that should fail with the code on `develop` but pass with this branch.  The test seems to represent reasonable parameter values. 

#### Documentation:

Added where it was missing.

#### Reviewer Suggestions: 

- The tests for F32 are done as far as I'm concerned and I was going to add parallel ones for grad_F32 and grad_2F1.  If you want anymore let me know
- The logic tracking minus signs seems bulky, suggestions appreciated but it is as hairy as it looks 
- F32 throws when the accumulator is infinite because that's a known non-convergence that sometimes occurs long before 

#### Copyright and Licensing

Copyright holder: Krzysztof Sakrejda

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

sakrejda commented:
@syclik I screwed up the doxygen file in here, I'll revert it and then point you to the problem file.

sakrejda commented:
@syclik Could you take a look at how to make this pass `make test-headers`?  It's `grad_reg_inc_beta` that triggers the failure  because it includes the `prim` version of `grad_2F1` and that function is `prim` so it only includes the `is_nan` from `prim` (and that fails when it gets a `var`.  If I included the `rev` version of `is_nan` in either `grad_reg_inc_beta` or `grad_2F1` it would pass, but both of those options seem wrong since `grad_2F1` calls `is_nan` but it's `prim` and `grad_reg_inc_beta` doesn't call `is_nan` ...

betanalpha commented on stan/math/prim/scal/fun/F32.hpp:
```diff
@@ -1,42 +1,89 @@
+     *   - <code>z</code> is less than 1
+     * This function will diverge if <code>z</code> is greater than 1.
+     * When <code>z</code> is 1, which is the case for the beta binomial
+     * cdf, it is hard to determine.
```
Like the other hypergeometric functions, 3F2 should converge at z = 1 subject to some conditions on the other inputs.  In particular, there should be a reflection symmetry that allows you to calculate near z = 1 by reflecting to 1 - z with corresponding changes to the other inputs.  This is critical for ensuring robust calculations -- see, for example, https://github.com/stan-dev/math/blob/develop/stan/math/prim/scal/fun/inc_beta_dda.hpp.

sakrejda replied:
👍 , I haven't fixed all the doc yet, I have a reference for when it does converge.  Do you think we should be checking for convergence conditions at the front? The original function didn't do this and I was aiming for more of a bugfix here with the intention of improving the functions as a second PR.  I'll either include the improvements here or make a specific second PR.

sakrejda replied:
I'm leaving the reflection symmetry for a subsequent issue #511

betanalpha commented on stan/math/prim/scal/fun/F32.hpp:
```diff
@@ -1,42 +1,89 @@
+        p = (a1 + k) / (b1 + k) * (a2 + k) / (b2 + k) * (a3 + k) / (k + 1);
 
-        tNew = exp(logT);
+        if (is_nan(p) || p == 0)
```
I'm a little iffy about this `is_nan` check.  When would p be NaN?  Poles, such as b1 = -k, can be checked beforehand and overflow can be avoided by computing and factoring partial terms is necessary.

sakrejda replied:
So if b1=-k or b2=-k this is triggered as the exit condition and I get answers within floating point precision of Mathematica as the answer.  I don't think we need an alternative calculation, although maybe checking for `b1=-k || b2=-k` would be clearer.  Suggestions?

sakrejda replied:
@betanalpha many sources seem to say the function is undefined when any denominator is a negative integer, mathematica gives an answer indistinguishable from terminating (as above) at the last defined term of the sum.  For this PR I'll either leave the code matching Mathematica or throw an error, I'll leave improving this situation for issue #511 as the current implementation doesn't do any better than this. Sound ok?

betanalpha commented on stan/math/prim/scal/fun/F32.hpp:
```diff
@@ -1,42 +1,89 @@
+      bool T_is_negative = false;
+      T p = 0.0;
+      do {
+        p = (a1 + k) / (b1 + k) * (a2 + k) / (b2 + k) * (a3 + k) / (k + 1);
```
This line definitely needs parentheses.  I much prefer the older version as it makes the possible poles much more explicit.

sakrejda replied:
This is a victim of too many git commits while looking for the bug, I'll revert this line to the original which was fine.

betanalpha commented on stan/math/prim/scal/fun/grad_2F1.hpp:
```diff
@@ -1,42 +1,71 @@
-      while (fabs(tDak * (a + (k - 1)) ) > precision || k == 0) {
-          const T r = ( (a + k) / (c + k) ) * ( (b + k) / (T)(k + 1) ) * z;
-          tDak = r * tDak * (a + (k - 1)) / (a + k);
+      do {
```
The do/while construction is somewhat awkward and certainly uncommon in C++.  If you're worried that the while loop needs to be executed at least once then you can just do a `while(true)` and break on the termination condition at the end of the loop.

sakrejda replied:
Seems really clear for algorithms that hit the loop at least once but I'm fine changing it.

betanalpha commented on stan/math/prim/scal/fun/grad_F32.hpp:
```diff
@@ -1,50 +1,86 @@
+      do {
+        p = (a1 + k) / (b1 + k) * (a2 + k) / (b2 + k) * (a3 + k) / (1 + k);
+
+        if (is_nan(p) || p == 0)
```
Same comment as above -- any pathological values should be traceable back to bad inputs which can be checked beforehand.  We want to be careful to distinguish between poles and values for which the power series just converges slowly.

sakrejda replied:
Ok, I can check inputs. This is only triggered when the denominator hits zero and it was an easy check.  If the series is converging slowly we only ever hit the iteration limit, not the `is_nan`check... but I can see how in the future this could cause subtle bugs if another source of `nan` appears.  I'll add input checks.

thelseraphim replied:
As far as making the intent more readable here, I think it might be nice to put the factor p (and input checks as michael mentioned) into a little inline function that gets called here, likewise the exp-log-sum-with-signs logic you're doing in the lines after.

sakrejda replied:
@thelseraphim I did break out convergence checking into an inline function but the exp-log-sum-with-signs logic carries state so I'm not sure how to break that out cleanly---you'd still have the state variable kicking around.

betanalpha commented on test/unit/math/prim/scal/fun/F32_test.cpp:
```diff
@@ -0,0 +1,54 @@
+              1e-6);
+
+  // This should terminate by zero numerator, no sign-flip
+  EXPECT_NEAR(11.28915378492300834453857665243661995978358572684,
```
Way more precision than necessary for an expert_near test with 1e-6 precision.

sakrejda replied:
How much precision should I be going for?

bob-carpenter commented:
C++ is left associative for / and *, so without parens, you get:

```
a / b * c / d * e = ((((a / b) * c) / d) * e)
```

If that's what you want, I'm OK without the parens.  The general principle is to assume the reader knows the language.

betanalpha commented:
It resolves to the right answer, it just toes the line of requiring
too much knowledge of operator precedence for my taste.  I’d
rather a bit more liberal use of parens and no ambiguity or
having to lookup precedence tables.

> On Feb 13, 2017, at 4:44 PM, Bob Carpenter <notifications@github.com> wrote:
> 
> C++ is left associative for / and *, so without parens, you get:
> 
> ```
> a / b * c / d * e = ((((a / b) * c) / d) * e)
> ```
> 
> If that's what you want, I'm OK without the parens. The general principle is to assume the reader knows the language.
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub <https://github.com/stan-dev/math/pull/488#issuecomment-279532279>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABdNlv-IUNsQNP9HXf-92RK-Xgf9DRj6ks5rcM6qgaJpZM4L96Ge>.
>

bob-carpenter commented:
I'll leave this one up to Michael and Krzysztof---a few extra parens aren't going to hurt anything.

The reason the general principle is in place is to stop people from documenting the language.  Without assuming the reader knows the language, it's impossible to decide where to stop documenting or where to stop adding parens. 

How do you feel about `a + b * c`?  Does that need parens or can readers be expected to know multiplication binds tighter than addition?  How about `a || b && c`?  Or an easily confusable constructor like `vector<int>(3, 7)`?  


> On Feb 13, 2017, at 4:48 PM, Michael Betancourt <notifications@github.com> wrote:
> 
> It resolves to the right answer, it just toes the line of requiring
> too much knowledge of operator precedence for my taste. I’d
> rather a bit more liberal use of parens and no ambiguity or
> having to lookup precedence tables.
> 
> > On Feb 13, 2017, at 4:44 PM, Bob Carpenter <notifications@github.com> wrote:
> > 
> > C++ is left associative for / and *, so without parens, you get:
> > 
> > ```
> > a / b * c / d * e = ((((a / b) * c) / d) * e)
> > ```
> > 
> > If that's what you want, I'm OK without the parens. The general principle is to assume the reader knows the language.
> > 
> > —
> > You are receiving this because your review was requested.
> > Reply to this email directly, view it on GitHub <https://github.com/stan-dev/math/pull/488#issuecomment-279532279>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABdNlv-IUNsQNP9HXf-92RK-Xgf9DRj6ks5rcM6qgaJpZM4L96Ge>.
> > 
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

sakrejda commented:
I had the parens in there because there are mixed types so arithmetic involves type conversion and I wanted to be clear about when the type conversion happens.  It's pretty clear here since it's always `x + k` so I'm fine dropping them.  It's expressions like `x/k*q` with `q` and `k` as `int`s that are more of a problem.

syclik commented:
I just added an include to the rev file that fails. I don't know of a
better way to do it at this moment. It over-includes in grad_inc_beta, but
it's necessary for the grad_2F1 header to compile under rev.


On Tue, Feb 14, 2017 at 9:11 AM, Krzysztof Sakrejda <
notifications@github.com> wrote:

> I had the parens in there because there are mixed types so arithmetic
> involves type conversion and I wanted to be clear about when the type
> conversion happens. It's pretty clear here since it's always x + k so I'm
> fine dropping them. It's expressions like x/k*q with q and k as ints that
> are more of a problem.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/488#issuecomment-279716813>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_FyeoDcMapco9Y7FzHEL6j-2HDHXvks5rcbYSgaJpZM4L96Ge>
> .
>

sakrejda commented:
@syclik Thanks! I prefer to have a limited number of people adding hacks so this is perfect.

bob-carpenter commented:
The EXPECT_NEAR argument will determine how much precision you need. This is the actual difference tolerance, not an absolute one, so it counts after the decimal point.

You never need more than 15 or 16 digits with double precision floating point, so you can treat that as a hard upper bound on the overall size, since everything after that will be ignored anyway.

> On Feb 14, 2017, at 9:04 AM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> @sakrejda commented on this pull request.
> 
> In test/unit/math/prim/scal/fun/F32_test.cpp:
> 
> > @@ -0,0 +1,54 @@
> +#include <gtest/gtest.h>
> +#include <stan/math/prim/scal/fun/F32.hpp>
> +
> +TEST(MathPrimScalFun, F32) {
> +  // This should terminate by convergence
> +  EXPECT_NEAR(2.5,
> +              stan::math::F32(1.0, 1.0, 1.0, 1.0, 1.0, 0.6),
> +              1e-6);
> +
> +  // This should terminate by zero numerator, no sign-flip
> +  EXPECT_NEAR(11.28915378492300834453857665243661995978358572684,
> 
> How much precision should I be going for?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
Yes, check as early as possible on the inputs if you can rather than producing an output and checking that.  The general principle is to fail as early as possible and provide a meaningful error message.

> On Feb 14, 2017, at 9:02 AM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> @sakrejda commented on this pull request.
> 
> In stan/math/prim/scal/fun/F32.hpp:
> 
> >  
> -        tNew = exp(logT);
> +        if (is_nan(p) || p == 0)
> 
> So if b1=-k or b2=-k this is triggered as the exit condition and I get answers within floating point precision of Mathematica as the answer. I don't think we need an alternative calculation, although maybe checking for b1=-k || b2=-k would be clearer. Suggestions?
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
If this is in the termination conditions of a loop, then you have to do the !is_nan check to avoid infinite loops unless you can guarantee you'll never get NaN inside the loop.

> On Feb 14, 2017, at 9:06 AM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> @sakrejda commented on this pull request.
> 
> In stan/math/prim/scal/fun/grad_F32.hpp:
> 
> > -        logT += (C > 0 ? 1 : -1) * log(fabs(C)) + logZ;
> -
> -        tNew = exp(logT);
> -
> -        gOld[0] = tNew * (gOld[0] / tOld + 1.0 / (a + k));
> -        gOld[1] = tNew * (gOld[1] / tOld + 1.0 / (b + k));
> -        gOld[2] = tNew * (gOld[2] / tOld + 1.0 / (c + k));
> -
> -        gOld[3] = tNew * (gOld[3] / tOld - 1.0 / (d + k));
> -        gOld[4] = tNew * (gOld[4] / tOld - 1.0 / (e + k));
> +      bool T_is_negative = false;
> +      T p = 0.0;
> +      do {
> +        p = (a1 + k) / (b1 + k) * (a2 + k) / (b2 + k) * (a3 + k) / (1 + k);
> +
> +        if (is_nan(p) || p == 0)
> 
> Ok, I can check inputs. This is only triggered when the denominator hits zero and it was an easy check. If the series is converging slowly we only ever hit the iteration limit, not the is_nancheck... but I can see how in the future this could cause subtle bugs if another source of nan appears. I'll add input checks.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
Not a problem either way with parens, but  `x / k * q` parses as `(x / k) * q` and integers are promoted when they're arithmetic operands and the other operand is `double` (or `var` or whatever).

> On Feb 14, 2017, at 9:11 AM, Krzysztof Sakrejda <notifications@github.com> wrote:
> 
> I had the parens in there because there are mixed types so arithmetic involves type conversion and I wanted to be clear about when the type conversion happens. It's pretty clear here since it's always x + k so I'm fine dropping them. It's expressions like x/k*q with q and k as ints that are more of a problem.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

bob-carpenter commented:
I don't know what happened here.  This pull request is touching hundreds of seemingly unrelated files and seems to be made up of a lot of unrelated commits.  Has it not been merged with develop yet or something?

sakrejda commented:
@bob-carpenter I thought I had branched it off develop correctly but maybe I had a stale develop.  It's obvious for me locally which files have changed.  I'll see if I can fix it...

sakrejda commented:
@bob-carpenter I have no idea why this looks like it does... if I do a `git pull origin develop` on my end it claims that it's up-to-date with develop.

sakrejda commented:
@bob-carpenter Sorry to keep at-ing you.  I checked that a fresh github-based compare of the branch this PR is based on vs. develop comes out clean (i.e.-only my dozen or so commits show up).  I think it's github not us, and specifically this PR not the branches that are messed up.

sakrejda commented:
Didn't we used to have a link here where one could go into jenkins and kill a build? I don't need this to run and I forgot to do the [skip ci] thing.

bob-carpenter commented:
Feel free to keep @-ing me, but I have no idea what's going on with GitHub, so you might want to @seantalts or @syclik

Any way you could split out the new files and make a new PR that would be possible to review?

syclik commented:
It looks like there is a legitimate merge conflict. Have you merged the
latest develop branch?

On Fri, Mar 17, 2017 at 7:29 PM, Bob Carpenter <notifications@github.com>
wrote:

> Feel free to keep @-ing me, but I have no idea what's going on with
> GitHub, so you might want to @seantalts or @syclik
>
> Any way you could split out the new files and make a new PR that would be
> possible to review?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/488#issuecomment-287495588>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAZ_FzIa2WoQwaQtp_9rtSdTY7jS8E6Iks5rmxdGgaJpZM4L96Ge>
> .
>

syclik commented:
maybe not. It looks like it's held up with cpplint.

seantalts commented:
I'm not sure about the link to kill a build, but can you click on the build "Details" and be brought to the Jenkins page and then kill it from there?

sakrejda commented:
@syclik I think I said this above but there's no (remaining) merge conflict.  Maybe it's screwed up because I merged in develop and fixed a merge conflict with develop at some point. For the next few days I'll just leave this PR so I can address the comments already on it but then I'll do a fresh PR---it looks like when I do that as a test the new PR doesn't have any of the messy history.

syclik commented:
Yup, the merge conflict isn't the problem now. It looks like cpplint to start.

bob-carpenter commented:
@sakrejda This is failing due to cpplint errors.

sakrejda commented:
Ok, it also lacks the tests I said I'd put in so it's not ready to merge
anyway.  Thanks.

On Mon, Mar 20, 2017, 5:24 PM Bob Carpenter <notifications@github.com>
wrote:

> @sakrejda <https://github.com/sakrejda> This is failing due to cpplint
> errors.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/math/pull/488#issuecomment-287902696>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAfA6ZGhHuCcVEQC-4O-MZP_a3Gv9Czaks5rnu58gaJpZM4L96Ge>
> .
>

sakrejda commented:
This now passes lint locally, I addressed the comments above except for the ones I punted to issue #511.  The tests that call the modified functions directly or indirectly have been run AFAIK and I'm waiting for the rest to complete.  The upshot was that I needed to move the gradient functions to log-scale power series as well and that lets me get EXPECT_NEAR to pass with 1e-8 error threshold while setting the precision to something like 1e-10.  Previously some of the old tests were passing but with a threshold of 1e-5/6 and some of the tests I added had errors up to 1e-4.  It seems like it's an open question what the default accuracy should be but now that it seems to achieve accuracy as requested I can look at some of the other gradient functions as a guide.  A second round of comments would be useful at this point (feel free to wait until the tests pass) for me to figure out what needs to happen to finish this off.  Keep in mind that I'd like to punt overall improvements to a separate issue and keep this to bug fixes and better defined behavior.

sakrejda commented:
@seantalts for some reason the Jenkins build fails and I can't even pull up the log, any chance you coudl look at it?

bob-carpenter commented:
Sean's out for a couple weeks.

I have no problem pulling up the log.  It fails in doxygen.  
error report below.

```
Started by upstream project "Math Pull Request" build number 677

originally caused by:
 GitHub pull request #488 of commit 8aa519fbd736e73c978970c9ad0a0563af0b5240, no merge conflicts.
[EnvInject] - Loading node environment variables.
Building remotely on 
gelman-group-mac
 (osx) in workspace /Users/Shared/Jenkins/gelman-group-mac/workspace/Math Pull Request - Doc - Doxygen
 > /usr/bin/git rev-parse --is-inside-work-tree # timeout=10
Fetching changes from the remote Git repository
 > /usr/bin/git config remote.origin.url 
https://github.com/stan-dev/math
 # timeout=10
Fetching upstream changes from 
https://github.com/stan-dev/math

 > /usr/bin/git --version # timeout=10
 > /usr/bin/git fetch --tags --progress 
https://github.com/stan-dev/math
 +refs/pull/*:refs/remotes/origin/pr/*
 > /usr/bin/git rev-parse refs/remotes/origin/pr/488/merge^{commit} # timeout=10
 > /usr/bin/git rev-parse refs/remotes/origin/origin/pr/488/merge^{commit} # timeout=10
Checking out Revision 2639e74dd2b9dc43df932c13987aa35f54777a57 (refs/remotes/origin/pr/488/merge)
 > /usr/bin/git config core.sparsecheckout # timeout=10
 > /usr/bin/git checkout -f 2639e74dd2b9dc43df932c13987aa35f54777a57
 > /usr/bin/git rev-list d02e54e592bc3f699b81d19f67bdb8eab2c9e6a3 # timeout=10
[Math Pull Request - Doc - Doxygen] $ /bin/sh -xe /var/folders/8k/9dmd0c0j0h58q068bszjgs2w0000gp/T/hudson1842211594034095582.sh
+ git clean -d -x -f
Removing doc/
Removing make/local
+ echo CC=clang++
+ echo O=0
[Math Pull Request - Doc - Doxygen] $ /bin/sh -xe /var/folders/8k/9dmd0c0j0h58q068bszjgs2w0000gp/T/hudson903899693017782406.sh
+ make clean-all
  removing test executables
rm -f -r doc/api
  removing dependency files
rm -f  
rm -f lib/cvodes_2.9.0/lib/libsundials_cvodes.a lib/cvodes_2.9.0/lib/libsundials_nvecserial.a lib/cvodes_2.9.0/src/cvodes/cvodea.o lib/cvodes_2.9.0/src/cvodes/cvodea_io.o lib/cvodes_2.9.0/src/cvodes/cvodes.o lib/cvodes_2.9.0/src/cvodes/cvodes_band.o lib/cvodes_2.9.0/src/cvodes/cvodes_bandpre.o lib/cvodes_2.9.0/src/cvodes/cvodes_bbdpre.o lib/cvodes_2.9.0/src/cvodes/cvodes_dense.o lib/cvodes_2.9.0/src/cvodes/cvodes_diag.o lib/cvodes_2.9.0/src/cvodes/cvodes_direct.o lib/cvodes_2.9.0/src/cvodes/cvodes_io.o lib/cvodes_2.9.0/src/cvodes/cvodes_sparse.o lib/cvodes_2.9.0/src/cvodes/cvodes_spbcgs.o lib/cvodes_2.9.0/src/cvodes/cvodes_spgmr.o lib/cvodes_2.9.0/src/cvodes/cvodes_spils.o lib/cvodes_2.9.0/src/cvodes/cvodes_sptfqmr.o lib/cvodes_2.9.0/src/nvec_ser/nvector_serial.o lib/cvodes_2.9.0/src/sundials/sundials_band.o lib/cvodes_2.9.0/src/sundials/sundials_dense.o lib/cvodes_2.9.0/src/sundials/sundials_direct.o lib/cvodes_2.9.0/src/sundials/sundials_iterative.o lib/cvodes_2.9.0/src/sundials/sundials_math.o lib/cvodes_2.9.0/src/sundials/sundials_nvector.o lib/cvodes_2.9.0/src/sundials/sundials_pcg.o lib/cvodes_2.9.0/src/sundials/sundials_sparse.o lib/cvodes_2.9.0/src/sundials/sundials_spbcgs.o lib/cvodes_2.9.0/src/sundials/sundials_spfgmr.o lib/cvodes_2.9.0/src/sundials/sundials_spgmr.o lib/cvodes_2.9.0/src/sundials/sundials_sptfqmr.o
  removing generated test files
rm -f 
+ make doxygen
mkdir -p doc/api
doxygen doxygen/doxygen.cfg
/Users/Shared/Jenkins/gelman-group-mac/workspace/Math Pull Request - Doc - Doxygen/stan/math/prim/scal/fun/grad_2F1.hpp:16: error: argument 'g' of command @param is not found in the argument list of stan::math::grad_2F1(T &g_a1, T &g_b1, const T &a1, const T &a2, const T &b1, const T &z, const T &precision=1e-10, int max_steps=1e5) (warning treated as error, aborting now)
make: *** [doxygen] Error 1
Build step 'Execute shell' marked build as failure
[WARNINGS] Skipping publisher since build result is FAILURE
Finished: FAILURE
```

sakrejda commented:
I have no idea why everyone else can see the Jenkins builds on this PR and I can't I'm going to close it and start a new one.
